tensorflow/core/common_runtime/dml/dml_device_context.h

```diff
@@ -52,6 +64,87 @@ class DMLDeviceContext : public DeviceContext {
   void CopyDeviceTensorToCPU(const Tensor* device_tensor, StringPiece edge_name,
                              Device* device, Tensor* cpu_tensor,
                              StatusCallback done) override;
+
+  // --------------------------------
+  // Allocation & Execution Helpers
+  // --------------------------------
+
+  // Initializes a given DML operator on the GPU. Note that this merely
+  // queues the initialization; the returned event will enter the signaled
+  // state when it completes. Note that we never supply any input bindings,
+  // because we never set DML_TENSOR_FLAG_OWNED_BY_DML .
+  DmlGpuEvent BindAndInitializeOperator(
+      IDMLOperatorInitializer* initializer,
+      Microsoft::WRL::ComPtr<IDMLBindingTable>&& binding_table,
+      ID3D12DescriptorHeap* heap_for_binding_table,
+      _In_opt_ const DML_BUFFER_BINDING* temporary_resource_binding,
+      _In_opt_ const DML_BUFFER_BINDING* persistent_resource_binding);
+
+  // Executes a DML operator. Note that this merely queues the execution; the
+  // returned event will enter the signaled state when it completes.
+  DmlGpuEvent BindAndExecuteOperator(
+      IDMLCompiledOperator* op,
+      Microsoft::WRL::ComPtr<IDMLBindingTable>&& binding_table,
+      ID3D12DescriptorHeap* heap_for_binding_table,
+      _In_opt_ const DML_BUFFER_BINDING* temporary_resource_binding,
```

Comment:
There seems to be a slightly inconsistent interface here: some methods take a D3D12BufferRegion (in which case the device context decides which barrier state it needs) whereas others take raw bindings/resources (in which case the caller is the one who has to figure out which state it wants). Same comment goes for the execution context. Should these be refactored to take either one or the other? (Deferring that change to a separate PR is also fine)

Reply:
The goal is to have 95% of the code use D3D12BufferRegion to avoid mistakes. I think I'll make the *Raw helpers on the execution context private and refactor these as well.